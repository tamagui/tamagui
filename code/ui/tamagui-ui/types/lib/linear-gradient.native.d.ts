export * from 'expo-linear-gradient';
//# sourceMappingURL=linear-gradient.native.d.ts.map