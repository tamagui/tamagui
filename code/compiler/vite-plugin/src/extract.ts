// fork from https://github.com/seek-oss/vanilla-extract

import type { TamaguiOptions } from '@tamagui/static'
import path from 'node:path'
import type { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import { normalizePath, type Environment } from 'vite'
import {
  Static,
  disableStatic,
  extractor,
  loadTamaguiBuildConfig,
  tamaguiOptions,
} from './loadTamagui'

import { createHash } from 'node:crypto'

export function tamaguiExtractPlugin(optionsIn?: Partial<TamaguiOptions>): Plugin {
  if (optionsIn?.disable) {
    return {
      name: 'tamagui-extract',
    }
  }

  const getHash = (input: string) => createHash('sha1').update(input).digest('base64')

  const clearCompilerCache = () => {
    memoryCache = {}
    cacheSize = 0
  }

  let memoryCache = {}
  let cacheSize = 0

  const cssMap = new Map<string, string>()

  let config: ResolvedConfig
  let server: ViteDevServer
  const virtualExt = `.tamagui.css`

  const getAbsoluteVirtualFileId = (filePath: string) => {
    if (filePath.startsWith(config.root)) {
      return filePath
    }
    return normalizePath(path.join(config.root, filePath))
  }

  function isVite6AndNotClient(environment?: Environment) {
    return environment?.name && environment.name !== 'client'
  }

  function isVite6Native(environment?: Environment) {
    return (
      environment?.name && (environment.name === 'ios' || environment.name === 'android')
    )
  }

  function invalidateModule(absoluteId: string) {
    if (!server) return

    const { moduleGraph } = server
    const modules = moduleGraph.getModulesByFile(absoluteId)

    if (modules) {
      for (const module of modules) {
        moduleGraph.invalidateModule(module)

        // Vite uses this timestamp to add `?t=` query string automatically for HMR.
        module.lastHMRTimestamp = module.lastInvalidationTimestamp || Date.now()
      }
    }
  }

  return {
    name: 'tamagui-extract',
    enforce: 'pre',

    configureServer(_server) {
      server = _server
    },

    async buildStart() {
      await loadTamaguiBuildConfig(optionsIn)
    },

    buildEnd() {
      extractor?.cleanupBeforeExit()
    },

    config(userConf) {
      userConf.optimizeDeps ||= {}
      userConf.optimizeDeps.include ||= []
      userConf.optimizeDeps.include.push('@tamagui/core/inject-styles')
    },

    async configResolved(resolvedConfig) {
      config = resolvedConfig
    },

    async resolveId(source) {
      if (isVite6Native(this.environment)) {
        return
      }

      if (
        tamaguiOptions?.disableServerOptimization &&
        isVite6AndNotClient(this.environment)
      ) {
        // only optimize on client - server should produce identical styles anyway!
        return
      }

      const [validId, query] = source.split('?')

      if (!validId.endsWith(virtualExt)) {
        return
      }

      // Absolute paths seem to occur often in monorepos, where files are
      // imported from outside the config root.
      const absoluteId = source.startsWith(config.root)
        ? source
        : getAbsoluteVirtualFileId(validId)

      // There should always be an entry in the `cssMap` here.
      // The only valid scenario for a missing one is if someone had written
      // a file in their app using the .tamagui.js/.tamagui.css extension
      if (cssMap.has(absoluteId)) {
        // Keep the original query string for HMR.
        return absoluteId + (query ? `?${query}` : '')
      }
    },

    /**
     * TODO
     *
     *   mainFields module:jsx breaks, so lets just have a mapping here
     *   where we load() and map it to the jsx path before transform
     *
     */

    async load(id) {
      if (disableStatic) {
        // only optimize on client - server should produce identical styles anyway!
        return
      }

      if (isVite6Native(this.environment)) {
        return
      }

      if (
        tamaguiOptions?.disableServerOptimization &&
        isVite6AndNotClient(this.environment)
      ) {
        return
      }

      const [validId] = id.split('?')
      return cssMap.get(validId)
    },

    transform: {
      order: 'pre',
      async handler(code, id) {
        if (disableStatic) {
          // only optimize on client - server should produce identical styles anyway!
          return
        }

        if (isVite6Native(this.environment)) {
          return
        }

        if (
          tamaguiOptions?.disableServerOptimization &&
          isVite6AndNotClient(this.environment)
        ) {
          return
        }

        const [validId] = id.split('?')
        if (!validId.endsWith('.tsx')) {
          return
        }

        const firstCommentIndex = code.indexOf('// ')
        const { shouldDisable, shouldPrintDebug } = Static!.getPragmaOptions({
          source: firstCommentIndex >= 0 ? code.slice(firstCommentIndex) : '',
          path: validId,
        })

        if (shouldPrintDebug) {
          console.trace(
            `Current file: ${id} in environment: ${this.environment?.name}, shouldDisable: ${shouldDisable}`
          )
          console.info(`\n\nOriginal source:\n${code}\n\n`)
        }

        if (shouldDisable) {
          return
        }

        const cacheEnv =
          this.environment.name === 'client' || this.environment.name === 'ssr'
            ? // same cache key for ssr and web since they are the same
              'web'
            : this.environment.name
        const cacheKey = getHash(`${cacheEnv}${code}${id}`)
        const cached = memoryCache[cacheKey]
        if (cached) {
          return cached
        }

        const extracted = await Static!.extractToClassNames({
          extractor: extractor!,
          source: code,
          sourcePath: validId,
          options: tamaguiOptions!,
          shouldPrintDebug,
        })

        if (!extracted) {
          return
        }

        const rootRelativeId = `${validId}${virtualExt}`
        const absoluteId = getAbsoluteVirtualFileId(rootRelativeId)

        let source = extracted.js

        if (extracted.styles) {
          this.addWatchFile(rootRelativeId)

          if (server && cssMap.has(absoluteId)) {
            invalidateModule(rootRelativeId)
          }

          source = `${source}\nimport "${rootRelativeId}";`
          cssMap.set(absoluteId, extracted.styles)
        }

        const codeOut = source.toString()
        const out = {
          code: codeOut,
          map: extracted.map,
        }

        cacheSize += codeOut.length
        // ~50Mb cache for recently compiler files
        if (cacheSize > 26214400) {
          clearCompilerCache()
        }
        memoryCache[cacheKey] = out

        return out
      },
    },
  }
}
