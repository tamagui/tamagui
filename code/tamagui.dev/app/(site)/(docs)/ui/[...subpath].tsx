import { ThemeTint } from '@tamagui/logo'
import { getMDXComponent } from 'mdx-bundler/client'
import React, { memo } from 'react'
import type { LoaderProps } from 'one'
import { useLoader } from 'one'
import { HeadInfo } from '~/components/HeadInfo'
import { DocsQuickNav } from '~/features/docs/DocsQuickNav'
import { MDXProvider } from '~/features/docs/MDXProvider'
import { MDXTabs } from '~/features/docs/MDXTabs'
import { useIsDocsTinted } from '~/features/docs/docsTint'
import { components } from '~/features/mdx/MDXComponents'
import { getOgUrl } from '~/features/site/getOgUrl'
import { nbspLastWord, SubTitle } from '../../../../components/SubTitle'
import { HomeH1 } from '../../../../features/site/home/HomeHeaders'

export async function generateStaticParams() {
  const { getAllFrontmatter } = await import('@tamagui/mdx-2')
  const frontmatters = getAllFrontmatter('data/docs/components')
  const paths = frontmatters.map((frontmatter) => {
    return {
      subpath: frontmatter.slug.replace('data/docs/components/', ''),
    }
  })

  const latestVersionPaths = paths.map((path) => {
    const parts = path.subpath.split('/')
    const withoutVersion = parts.slice(0, parts.length - 1)
    return withoutVersion.join('/')
  })

  const deduped = [...new Set(latestVersionPaths)].map((subpath) => ({ subpath }))

  const allPaths = [...paths, ...deduped]

  return allPaths
}

export async function loader(props: LoaderProps) {
  const { getMDXBySlug, getAllVersionsFromPath } = await import('@tamagui/mdx-2')

  const subpath = Array.isArray(props.params.subpath)
    ? props.params.subpath[0]
    : props.params.subpath

  const { frontmatter, code } = await getMDXBySlug('data/docs/components', subpath)
  const [componentName, componentVersion] = subpath.split('/')
  const versions = getAllVersionsFromPath(`data/docs/components/${componentName}`)
  return {
    frontmatter: {
      ...frontmatter,
      version: componentVersion || versions[0],
      versions: versions,
    },
    code,
  }
}

export function DocComponentsPage() {
  const { frontmatter, code } = useLoader(loader)
  const Component = React.useMemo(() => getMDXComponent(code), [code])

  // useEffect(() => {
  //   const url = new URL(location.href)
  //   url.pathname = `${pathname}/${frontmatter.version}`
  //   if (Array.isArray(params.subpath)) {
  //     url.pathname = url.pathname.replace('[...subpath]', params.subpath[0])
  //   }
  //   router.replace(url)
  // }, [])

  return (
    <>
      <HeadInfo
        title={`${frontmatter.title} | Tamagui â€” style library and UI kit for React`}
        description={frontmatter.description}
        openGraph={{
          images: [
            {
              url:
                frontmatter.image ??
                getOgUrl({
                  type: 'component',
                  title: frontmatter.title,
                  demoName: frontmatter.demoName ?? undefined,
                  description: frontmatter.description ?? '',
                }),
              width: 1200,
              height: 630,
            },
          ],
        }}
      />
      {/* {frontmatter.version !== frontmatter.versions?.[0] && (
        <OldVersionNote
          name={frontmatter.title}
          href={`/primitives/docs/components/${frontmatter.subpath.replace(frontmatter.version, '')}`}
        />
      )} */}

      <HomeH1>{nbspLastWord(frontmatter.title)}</HomeH1>

      <SubTitle>{nbspLastWord(frontmatter.description || '')}</SubTitle>

      <MDXProvider frontmatter={frontmatter}>
        <DocsThemeTint>
          <MDXTabs id="type" defaultValue="styled">
            <Component components={components as any} />
          </MDXTabs>
        </DocsThemeTint>
      </MDXProvider>

      <DocsQuickNav key={frontmatter.slug} />
    </>
  )
}

const DocsThemeTint = memo(({ children }: { children: any }) => {
  const isTinted = useIsDocsTinted()
  return <ThemeTint disable={!isTinted}>{children}</ThemeTint>
})
