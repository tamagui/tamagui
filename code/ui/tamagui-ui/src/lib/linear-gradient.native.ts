export * from 'expo-linear-gradient'
