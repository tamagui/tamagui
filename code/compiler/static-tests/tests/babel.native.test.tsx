import * as React from 'react'
import { expect, test } from 'vitest'

import { extractForNative } from './lib/extract'

process.env.TAMAGUI_TARGET = 'native'
process.env.IS_STATIC = ''

window['React'] = React

test('basic extraction', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    export function Test() {
      return (
        <YStack backgroundColor="red" />
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('theme value extraction should work when no theme variables used', async () => {
  // here we override default "$color" so it should flatten totally
  // actually since we do unstyled: false and it expands to $color it acceses it
  // were not smart enough yet to detect its later overriden :/
  // that could be a perf optimization, but also have work to improve flattening soon anyway
  const output = await extractForNative(`
    import { Paragraph } from '@tamagui/ui'
    export function Test() {
      return (
        <Paragraph color="red">hello world</Paragraph>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('theme value extraction should work when theme variables used', async () => {
  const output = await extractForNative(`
    import { Paragraph } from '@tamagui/ui'
    export function Test() {
      return (
        <Paragraph>hello world</Paragraph>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('basic conditional extraction', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    let x = true
    export function Test() {
      return (
        <>
          <YStack backgroundColor={x ? 'red' : 'blue'} />
          <YStack {...x && { backgroundColor: 'red' }} />
        </>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('flat transform props', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    export function Test(isLoading) {
      return (
        <YStack
          scale={isLoading ? 1 : 2}
          x={10}
          y={20}
          rotate="10deg"
        />
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('handles style order merge properly', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    export function Test(props) {
      return (
        <YStack
          scale={props.isLoading ? 1 : 2}
          x={10}
          {...props}
          rotate="10deg"
        />
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test(`normalize ternaries flips the conditional properly`, async () => {
  const inputCode = `
  import { Stack } from '@tamagui/ui'
  export function Test(props) {
    return (
      <Stack marginBottom={props !== 123 ? 12 : 0} />
    )
  }
`
  const output = await extractForNative(inputCode)
  const outCode = output?.code ?? ''
  expect(outCode).toContain(`props === 123 ? _sheet["0"] : _sheet["1"]`)
  expect(outCode).toMatchSnapshot()
})

test(`normalize ternaries with the conditional dynamic values`, async () => {
  const inputCode = `
  import { Stack } from '@tamagui/ui'
  export function Test(props) {
    return (
      <Stack marginBottom={props !== 123 ? 12 : props.mb} />
    )
  }
`
  const output = await extractForNative(inputCode)
  const outCode = output?.code ?? ''
  expect(outCode).toContain(`props !== 123 ? 12 : props.mb`)
  expect(outCode).toMatchSnapshot()
})

test('normalize dynamic values with no theme access', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    const height = 200
    export function Test(props) {
      return (
        <YStack height={height} width={props.width}/>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('normalize dynamic values with theme access only', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    export function Test(props) {
      return (
        <YStack bg='$color'/>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('do NOT flatten dynamic values with theme access', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    export function Test(props) {
      return (
        <YStack bg='$color' height={props.height}/>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('do NOT flatten dynamic values with theme access, dynamic values, and conditional', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    export function Test(props) {
      return (
        <YStack bg={props.isLoading ? '$color' : 'red'} height={props.height}/>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})

test('do NOT flatten multiple dynamic values with theme access and conditional', async () => {
  const output = await extractForNative(`
    import { YStack } from '@tamagui/ui'
    export function Test(props) {
      return (
        <YStack bg={props.isLoading ? '$color' : 'red'} height={props.height} width={props.width}/>
      )
    }
  `)
  const code = output?.code ?? ''
  expect(code).toMatchSnapshot()
})
