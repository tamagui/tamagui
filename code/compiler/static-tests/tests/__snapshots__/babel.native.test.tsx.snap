// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic conditional extraction 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "flexDirection": "column"
  },
  "1": {
    "backgroundColor": "red"
  },
  "2": {
    "backgroundColor": "blue"
  },
  "3": {
    "flexDirection": "column"
  },
  "4": {
    "backgroundColor": "red"
  },
  "5": {}
});
const __ReactUseMemo = require('react').useMemo;
const __withStableStyle = require('@tamagui/core')._withStableStyle;
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
let x = true;
export function Test() {
  return <>
          <_ReactNativeViewWrapper expressions={[x]} />
          <_ReactNativeViewWrapper2 expressions={[x]} />
        </>;
}
const _ReactNativeViewWrapper = __withStableStyle(__ReactNativeView, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["0"], _expressions[0] ? _sheet["1"] : _sheet["2"]];
  }, [..._expressions]);
});
const _ReactNativeViewWrapper2 = __withStableStyle(__ReactNativeView, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["3"], _expressions[0] ? _sheet["4"] : _sheet["5"]];
  }, [..._expressions]);
});"
`;

exports[`basic extraction 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "flexDirection": "column",
    "backgroundColor": "red"
  }
});
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test() {
  return <__ReactNativeView style={_sheet["0"]} />;
}"
`;

exports[`do NOT flatten dynamic values with theme access 1`] = `
"const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test(props) {
  return <YStack bg='$color' height={props.height} />;
}"
`;

exports[`do NOT flatten dynamic values with theme access, dynamic values, and conditional 1`] = `
"const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test(props) {
  return <YStack bg={props.isLoading ? '$color' : 'red'} height={props.height} />;
}"
`;

exports[`do NOT flatten multiple dynamic values with theme access and conditional 1`] = `
"const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test(props) {
  return <YStack bg={props.isLoading ? '$color' : 'red'} height={props.height} width={props.width} />;
}"
`;

exports[`flat transform props 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "flexDirection": "column",
    "transform": [{
      "translateY": 20
    }, {
      "translateX": 10
    }, {
      "rotate": "10deg"
    }]
  },
  "1": {
    "transform": [{
      "scale": 1
    }]
  },
  "2": {
    "transform": [{
      "scale": 2
    }]
  }
});
const __ReactUseMemo = require('react').useMemo;
const __withStableStyle = require('@tamagui/core')._withStableStyle;
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test(isLoading) {
  return <_ReactNativeViewWrapper expressions={[isLoading]} />;
}
const _ReactNativeViewWrapper = __withStableStyle(__ReactNativeView, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["0"], _expressions[0] ? _sheet["1"] : _sheet["2"]];
  }, [..._expressions]);
});"
`;

exports[`handles style order merge properly 1`] = `
"const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test(props) {
  return <YStack scale={props.isLoading ? 1 : 2} x={10} {...props} rotate="10deg" />;
}"
`;

exports[`normalize dynamic values with no theme access 1`] = `
"const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
const height = 200;
export function Test(props) {
  return <YStack height={height} width={props.width} />;
}"
`;

exports[`normalize dynamic values with theme access only 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "flexDirection": "column"
  }
});
const __ReactUseMemo = require('react').useMemo;
const __withStableStyle = require('@tamagui/core')._withStableStyle;
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { YStack } from 'tamagui';
export function Test(props) {
  return <_ReactNativeViewWrapper />;
}
const _ReactNativeViewWrapper = __withStableStyle(__ReactNativeView, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["0"], {
      "backgroundColor": theme.color.get()
    }];
  }, [..._expressions]);
});"
`;

exports[`normalize ternaries flips the conditional properly 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "marginBottom": 0
  },
  "1": {
    "marginBottom": 12
  }
});
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { Stack } from 'tamagui';
export function Test(props) {
  return <__ReactNativeView style={props === 123 ? _sheet["0"] : _sheet["1"]} />;
}"
`;

exports[`normalize ternaries with the conditional dynamic values 1`] = `
"const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { Stack } from 'tamagui';
export function Test(props) {
  return <Stack marginBottom={props !== 123 ? 12 : props.mb} />;
}"
`;

exports[`theme value extraction should work when no theme variables used 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "fontFamily": "Inter",
    "letterSpacing": 0,
    "fontSize": 14,
    "lineHeight": 23,
    "color": "red"
  }
});
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { Paragraph } from 'tamagui';
export function Test() {
  return <__ReactNativeText style={_sheet["0"]}>hello world</__ReactNativeText>;
}"
`;

exports[`theme value extraction should work when theme variables used 1`] = `
"const __ReactNativeStyleSheet = require('react-native').StyleSheet;
const _sheet = __ReactNativeStyleSheet.create({
  "0": {
    "fontFamily": "Inter",
    "letterSpacing": 0,
    "fontSize": 14,
    "lineHeight": 23
  }
});
const __ReactUseMemo = require('react').useMemo;
const __withStableStyle = require('@tamagui/core')._withStableStyle;
const __ReactNativeView = require('react-native').View;
const __ReactNativeText = require('react-native').Text;
import { Paragraph } from 'tamagui';
export function Test() {
  return <_ReactNativeTextWrapper>hello world</_ReactNativeTextWrapper>;
}
const _ReactNativeTextWrapper = __withStableStyle(__ReactNativeText, (theme, _expressions) => {
  return __ReactUseMemo(() => {
    return [_sheet["0"], {
      "color": theme.color.get()
    }];
  }, [..._expressions]);
});"
`;
